includes/class-wc-admin-notes-order-milestones.php

```diff
@@ -0,0 +1,251 @@
+
+	/**
+	 * Hook to process order milestones.
+	 */
+	const PROCESS_ORDERS_MILESTONE_HOOK = 'wc_admin_process_orders_milestone';
+
+	/**
+	 * Allowed order statuses for calculating milestones.
+	 *
+	 * @var array
+	 */
+	protected $allowed_statuses = array(
+		'pending',
+		'processing',
+		'completed',
+	);
+
+	/**
+	 * Orders count cache.
+	 *
+	 * @var int
+	 */
+	protected $orders_count = null;
+
+	/**
+	 * Further order milestone thresholds.
+	 *
+	 * @var array
+	 */
+	protected $milestones = array(
+		100,
+		250,
+		500,
+		1000,
+		5000,
+		10000,
+		500000,
+		1000000,
+	);
+
+	/**
+	 * Delay hook attachment until after the WC post types have been registered.
+	 *
+	 * This is required for retrieving the order count.
+	 */
+	public function __construct() {
+		add_action( 'woocommerce_after_register_post_type', array( $this, 'init' ) );
+	}
+
+	/**
+	 * Hook everything up.
+	 */
+	public function init() {
+		if ( ! wp_next_scheduled( self::PROCESS_ORDERS_MILESTONE_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::PROCESS_ORDERS_MILESTONE_HOOK );
+		}
+
+		add_action( 'wc_admin_installed', array( $this, 'backfill_last_milestone' ) );
+
+		if ( 10 <= $this->get_orders_count() ) {
+			add_action( 'woocommerce_new_order', array( $this, 'first_two_milestones' ) );
+		}
+
+		add_action( self::PROCESS_ORDERS_MILESTONE_HOOK, array( $this, 'other_milestones' ) );
+	}
+
+	/**
+	 * Get the total count of orders (in the allowed statuses).
+	 *
+	 * @return int Total orders count.
+	 */
+	public function get_orders_count() {
+		if ( is_null( $this->orders_count ) ) {
+			$status_counts       = array_map( 'wc_orders_count', $this->allowed_statuses );
+			$this->orders_count  = array_sum( $status_counts );
+		}
+
+		return $this->orders_count;
+	}
+
+	/**
+	 * Add a milestone notes for the store's first order and first 10 orders.
+	 *
+	 * @param int $order_id WC_Order ID.
+	 */
+	public function first_two_milestones( $order_id ) {
+		$order = wc_get_order( $order_id );
+
+		// Make sure this is the first pending/processing/completed order.
+		if ( ! in_array( $order->get_status(), $this->allowed_statuses ) ) {
+			return;
+		}
+
+		$orders_count = $this->get_orders_count();
+
+		if ( 1 === $orders_count ) {
+			// Add the first order note.
+			$note = new WC_Admin_Note();
+			$note->set_title( __( 'First order', 'woocommerce-admin' ) );
+			$note->set_content(
+				__( 'Congratulations on getting your first order from a customer! Learn how to manage your orders.', 'woocommerce-admin' )
+			);
+			$note->set_type( WC_Admin_Note::E_WC_ADMIN_NOTE_INFORMATIONAL );
+			$note->set_icon( 'trophy' );
+			$note->set_name( self::FIRST_ORDER_NOTE_NAME );
+			$note->set_source( 'woocommerce-admin' );
+			$note->add_action( 'learn-more', __( 'Learn more', 'woocommerce-admin' ), 'https://docs.woocommerce.com/document/managing-orders/' );
+			$note->save();
+		}
+
+		if ( 10 === $orders_count ) {
+			// Add the first ten orders note.
```

Comment:
Should the first order milestone be deleted here too, akin to what is done on the other milestones?

`WC_Admin_Notes::delete_notes_with_name( self::ORDERS_MILESTONE_NOTE_NAME );`

Reply:
I'm not sure. Let's ask @MarinaWoo - if a store hits the 10 order milestone (or even the 100) quickly, should only one milestone alert display, or should the first order and 10 order milestones be in the inbox alongside each other?

Reply:
It is all so very relative, it is hard to know! I think if someone is tinkering and testing they would be aware that they are 'forcing' the milestones, and therefore if they showed up fast it would be okay. I do think 'deleted' orders should (if possible) be included in the count towards milestones so as to avoid the scenario @timmyc is talking about.

Reply:
As it currently sits, deleted orders would only reset the notifications for the first order and the tenth order. It wouldn't rollback any of the 100+ milestones.